### PR TITLE
[SPARK-28123][SQL] String Functions: support btrim

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
Spark support `trim`/`ltrim`/`rtrim` now. The function `btrim` is an alternate form of `TRIM(BOTH <chars> FROM <expr>)`.
`btrim` removes the longest string consisting only of specified characters from the start and end of a string.

The mainstream database support this feature show below:

**Postgresql**
https://www.postgresql.org/docs/11/functions-binarystring.html

**Vertica**
https://www.vertica.com/docs/9.2.x/HTML/Content/Authoring/SQLReferenceManual/Functions/String/BTRIM.htm?tocpath=SQL%20Reference%20Manual%7CSQL%20Functions%7CString%20Functions%7C_____5

**Redshift**
https://docs.aws.amazon.com/redshift/latest/dg/r_BTRIM.html

**Druid**
https://druid.apache.org/docs/latest/querying/sql.html#string-functions

**Greenplum**
http://docs.greenplum.org/6-8/ref_guide/function-summary.html

### Why are the changes needed?
btrim is very useful.

### Does this PR introduce _any_ user-facing change?
Yes. btrim is a new function

### How was this patch tested?
Jenkins test.

### CODE REVIEW

**Comment by beliefer** [comment]
retest this please

**Comment by beliefer** [comment]
cc @cloud-fan

File: sql/core/src/test/resources/sql-tests/inputs/postgreSQL/strings.sql [hunk 1]
```diff
@@ -647,7 +647,9 @@ SELECT repeat('Pg', 4);
 SELECT repeat('Pg', -4);
 
 SELECT trim(binary('\\000') from binary('\\000Tom\\000'));
--- [SPARK-28123] Add support btrim
+SELECT btrim(binary('\\000trim\\000'), binary('\\000'));
+SELECT btrim(binary(''), binary('\\000'));
+SELECT btrim(binary('\\000trim\\000'), binary(''));
 -- SELECT btrim(E'\\000trim\\000'::bytea, E'\\000'::bytea);
 -- SELECT btrim(''::bytea, E'\\000'::bytea);
 -- SELECT btrim(E'\\000trim\\000'::bytea, ''::bytea);
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/postgreSQL/strings.sql:655]: shall we remove the above 3 lines?

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/test/resources/sql-tests/inputs/postgreSQL/strings.sql (hunk 1 above)
Comment (beliefer) [reply sql/core/src/test/resources/sql-tests/inputs/postgreSQL/strings.sql:655]: OK

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by beliefer** [comment]
@cloud-fan Thanks for your work!

**Comment by gengliangwang** [comment]
As per https://www.postgresql.org/docs/11/functions-binarystring.html, the result type of `btrim` should be of binary type instead of string type.
@beliefer cloud you fix it? Or, can I revert it first(this PR blocks https://github.com/apache/spark/pull/31349)?

**Comment by beliefer** [comment]
@btrim means trim from leading and trailing. not binary

**Comment by gengliangwang** [comment]
@beliefer From the first link you provide in PR description:
> https://www.postgresql.org/docs/11/functions-binarystring.html

I thought that the function is for binary input. Sorry for the misunderstanding.
So in PostgreSQL the behavior is 
```
btrim(String) => String
btrim(Binary) => Binary
```
In current Spark
```
btrim(String) => String
btrim(Binary) => String
```

**Comment by beliefer** [comment]
@gengliangwang Let us ignore the return type of btrim. In Postgresql, Vertica, Redshift, Druid, Greenplum, btrim is used to remove the leading and trailing space characters from `str`.

The behavior of Postgresql just returns Binary

**Comment by gengliangwang** [comment]
> I created #31080 before this PR to implement the behavior that is returns Binary. But that was a mistake.

I don't think it's a total mistake. Is there any existing database that has the behavior?
```
btrim(Binary) => String
```

I checked PostgreSQL and Google Bigquery and they all return **Binary** type for btrim(Binary). It seems more reasonable for trim/btrim functions: if a users wants to trim a binary column, he/she has to run `cast(trim(binary_col) as binary)` and Spark converts binary to string and then convert the trimmed result as binary.

**Comment by beliefer** [comment]
@gengliangwang Regardless of the data type, the databases listed above have consistent behavior trim from leading and trailing.
Although `btrim(Binary) => Binary` in PostgreSQL, but Vertica, Redshift and Druid follows `btrim(String) => String`. So we should think about wheather or not support the implict transform from Binary to String

**Comment by cloud-fan** [comment]
does any database do `trim(binary) => string` except Spark?

**Comment by beliefer** [comment]
@cloud-fan @gengliangwang 
There not any database do `trim(binary) => string`, so I think should remove the `ImplicitCastInputTypes` from `String2TrimExpression`.

**Comment by gengliangwang** [comment]
> But it is a break change.

Hmm, then I prefer keeping the current behavior: Trimming binary input is allowed in non-ansi mode, while disallowed in ANSI mode.

**Comment by beliefer** [comment]
> > But it is a break change.
> 
> Hmm, then I prefer keeping the current behavior: Trimming binary input is allowed in non-ansi mode, while disallowed in ANSI mode.

OK. Keep the current behavior.

**Pull request closed without merging**